doc/source/tutorials/filters.rst

```diff
@@ -0,0 +1,372 @@
+future.
+
+Compression filters
+~~~~~~~~~~~~~~~~~~~
+
+There are several filters performing generic compression, which are the following:
+
+* ``TILEDB_FILTER_GZIP``: Compresses with `Gzip <http://www.zlib.net/>`__
+* ``TILEDB_FILTER_ZSTD``: Compresses with `Zstandard <http://facebook.github.io/zstd/>`__
+* ``TILEDB_FILTER_LZ4``: Compresses with `LZ4 <https://github.com/lz4/lz4>`__
+* ``TILEDB_FILTER_RLE``: Compresses with `run-length encoding <https://en.wikipedia.org/wiki/Run-length_encoding>`__
+* ``TILEDB_FILTER_BZIP2``: Compresses with `Bzip2 <http://www.bzip.org/>`__
+* ``TILEDB_FILTER_DOUBLE_DELTA``: Compresses with double-delta encoding
+
+All of these filters support one filter option to set the compression level,
+although some compressors such as RLE currently ignore the setting. The filter
+option is:
+
+* ``TILEDB_COMPRESSION_LEVEL`` (type ``int32_t``): The compression level to
+  use. Default: -1 (compressor-specific default).
+
+Byteshuffle
+~~~~~~~~~~~
+
+The filter ``TILEDB_FILTER_BYTESHUFFLE`` performs byte shuffling of data as a
+way to improve compression ratios. The
+`byte shuffle implementation <{tiledb_src_root_url}/external/include/blosc/shuffle.h>`_
+used by TileDB comes from the `Blosc <blosc.org>`_ project.
+
+The byte shuffling process rearranges the bytes of the input attribute cell
+values in a deterministic and reversible manner designed to result in long runs
+of similar bytes that can be compressed more effectively by a generic compressor
+than the original unshuffled elements. Typically this filter would not be used
+on its own, but rather immediately followed by a compression filter in a filter
+list.
+
+For example, consider three 32-bit unsigned integer values ``1, 2, 3``, which
+have the following little-endian representation when stored adjacent in memory:
+
+.. code-block:: none
+
+    0x01 0x00 0x00 0x00 0x02 0x00 0x00 0x00 0x03 0x00 0x00 0x00
+
+The byte shuffle operation will rearrange the bytes of these integer elements in
+memory such that the resulting array of bytes will contain each element's first
+byte, followed by each element's second byte, etc. After shuffling the bytes
+would therefore be:
+
+.. code-block:: none
+
+    0x01 0x02 0x03 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00 0x00
+
+Note the longer run of zero-valued bytes, which will compress more efficiently.
+
+The byteshuffle filter does not support any options.
+
+Bitshuffle
+~~~~~~~~~~
+
+The filter ``TILEDB_FILTER_BITSHUFFLE`` performs
+`bit shuffling <https://www.sciencedirect.com/science/article/pii/S2213133715000694>`_
+of data as a way to improve compression ratios. The bitshuffle implementation
+used in TileDB comes from `<https://github.com/kiyo-masui/bitshuffle>`_.
+
+Bitshuffling is conceptually very similar to byteshuffling, but operates on the
+bit granularity rather than the byte granularity. Shuffling at the bit level
+can increase compression ratios even further than the byteshuffle filter.
```

Comment:
maybe add a note here about the tradeoff: "... at the expense of increased computation" or something like that